File: cpp/mugwump.cpp
#include <iostream>
#include <cmath>

const int M = 4;
const int S = 10;
const int G = 10;
struct Pos { int x, y; };
Pos m[M];
bool caught[M];

void placeMonsters()
{
    for (int i = 0 ; i < M ; ++i)
    {
        bool again = true;
        while (again)
        {
            m[i].x = rand() % S;
            m[i].y = rand() % S;
            again = false;
            for (int j = 0 ; j < i ; ++j)
            {
                if (m[i].x == m[j].x && m[i].y == m[j].y)
                {
                    again = true;
                }
            }
        }
    }
}

void printMonsters()
{
    std::cout << "Монстры:\n";
    for (int i = 0 ; i < M ; ++i)
    {
        std::cout << i << ": ";
        if (caught[i])
        {
            std::cout << "пойман\n";
        }
        else
        {
            std::cout << m[i].x << " " << m[i].y << "\n";
        }
    }
}

Pos inputMove()
{
    int x, y;
    do
    {
        std::cout << "Введите координаты: ";
        std::cin >> x >> y;
    }
    while (x < 0 || x >= S
        || y < 0 || y >= S);
    return Pos{x, y};
}

void checkDistance(Pos pos)
{
    for (int i = 0 ; i < M ; ++i)
    {
        if (caught[i])
        {
            continue;
        }
        std::cout << "Монстр " << i + 1;
        if (pos.x == m[i].x && pos.y == m[i].y)
        {
            caught[i] = true;
            std::cout << " пойман!\n";
        }
        else
        {
            int dx = pos.x - m[i].x;
            int dy = pos.y - m[i].y;
            double dist = std::sqrt(dx * dx + dy * dy);
            std::cout << " на расстоянии " << dist << "\n";
        }
    }
}

bool allCaught()
{
    for (int i = 0 ; i < M ; ++i)
    {
        if (!caught[i])
        {
            return false;
        }
    }
    return true;
}

void finalMessage(int move)
{
    if (move <= G)
    {
        std::cout << "Поздравляем, вы поймали всех монстров за " << move << " ходов!\n";
    }
    else
    {
        std::cout << "Вы проиграли, можете попробовать снова.\n";
    }
}

int main()
{
    srand((int)time(NULL));

    int move;
    placeMonsters();
    for (move = 1 ; move <= G ; ++move)
    {
        std::cout << "Ход " << move << "\n";
        //printMonsters();
        checkDistance(inputMove());
        if (allCaught())
        {
            break;
        }
    }
    finalMessage(move);
}

File: python/mugwump.py
#!/usr/bin/python3
import random
from dataclasses import dataclass
import math

@dataclass
class Pos:
    x: int
    y: int

M = 4
S = 10
G = 10
m = [Pos(0, 0)] * M
caught = [False] * M

def printMonsters():
    print('Монстры:')
    for i in range(M):
        print(f'{i}: ', end='')
        if caught[i]:
            print('пойман')
        else:
            print(f'{m[i].x} {m[i].y}')

def placeMonsters():
    for i in range(M):
        again = True
        while again:
            m[i] = Pos(random.randint(0, S - 1), \
                    random.randint(0, S - 1))
            again = False
            for j in range(i):
                if m[i].x == m[j].x and m[i].y == m[j].y:
                    again = True

def inputMove():
    x = -1
    y = -1
    while x < 0 or x >= S or y < 0 or y >= S:
        x, y = [int(s) for s in input('Введите координаты: ').split()]
    return Pos(x, y)

def checkDistance(pos):
    for i in range(M):
        if caught[i]:
            continue
        print(f'Монстр {i + 1}', end='')
        if pos.x == m[i].x and pos.y == m[i].y:
            caught[i] = True;
            print(' пойман!')
        else:
            dx = pos.x - m[i].x
            dy = pos.y - m[i].y
            dist = math.sqrt(dx * dx + dy * dy)
            print(f' на расстоянии {dist:.3}')

def allCaught():
    for c in caught:
        if not c:
            return False
    return True

def finalMessage(move):
    if move <= G:
        print(f'Поздравляем, вы поймали всех монстров за {move} ходов!')
    else:
        print('Вы проиграли, можете попробовать снова.')

placeMonsters()
move = 1
while move <= G:
    print(f'Ход {move}')
    #printMonsters()
    checkDistance(inputMove())
    if allCaught():
        break
    move += 1
finalMessage(move)
